src/lib/ui/Menu/Menu.tsx

```diff
@@ -0,0 +1,179 @@
+import classNames from 'classnames'
+import {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useState,
+} from 'react'
+import { createPortal } from 'react-dom'
+
+export interface MenuProps {
+  children: JSX.Element | JSX.Element[]
+  verticalOffset?: number
+  horizontalOffset?: number
+  edgeOffset?: number
+  button: (props: {
+    open: (event: React.MouseEvent<HTMLElement>) => void
+  }) => React.ReactElement
+  BackgroundProps?: Partial<{
+    className?: string
+  }>
+}
+
+interface MenuContextProps {
+  close: () => void
+}
+
+const MenuContext = createContext<MenuContextProps | undefined>(undefined)
+
+const Menu = ({
+  verticalOffset = 5,
+  horizontalOffset = 0,
+  edgeOffset = 5,
+  children,
+  button,
+  BackgroundProps,
+}: MenuProps) => {
+  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null)
+  const [documentEl, setDocumentEl] = useState<null | Element>(null)
+  const [contentEl, setContentEl] = useState<HTMLDivElement | null>(null)
```

Comment:
@mikewuu Not saying useState is wrong here, but I feel like `useRef` is normally used when dealing with dom element refs (because their lifecycle is outside the react tree lifecycle maybe?)

Reply:
>  lifecycle is outside the react tree lifecycle

Yep. In this case I wanted it to be inside the lifecycle, so we know when to re-calculate.

For side-effects, I'd go with `useRef`.